src/next/useSearchParams.ts

```diff
@@ -1,63 +1,66 @@
-import { useCallback, useEffect, useMemo, useRef } from "react"
-import {
-  usePathname,
-  useSearchParams as useNextSearchParams,
-  useRouter
-} from "next/navigation"
+import { useCallback, useEffect, useRef } from "react"
+import { useSearchParams as useNextSearchParams } from "next/navigation"
 
 type SearchParamsSetterValue =
   | URLSearchParams
   | ((prevSearchParams: URLSearchParams) => URLSearchParams)
 
 type SetSearchParams = (newSearchParams: SearchParamsSetterValue) => void
 
-const useSearchParams = (): [URLSearchParams, SetSearchParams] => {
-  const pathname = usePathname()
-  const router = useRouter()
-  const search = useNextSearchParams()
-
+/**
+ * Returns a function for setting search params in the browser URL.
+ *  - the first call within a single render cycle will add a new entry in the
+ *   history stack via pushState
+ *  - subsequent calls will replace the previous entry in the history stack
+ *    and will receive the previously updated value in setter function
+ *
+ * NOTE: This is intended for use with NextJS, but could work with any framework
+ * that supports direct usage of window.history.pushState / replaceState.
+ */
+const useSetSearchParams = (): SetSearchParams => {
   /**
    * Keep track of whether navigate has been called in the current render cycle
    * to avoid adding extra entries in the history stack.
    */
   const hasNavigatedRef = useRef(false)
-  const searchParams = useMemo(() => new URLSearchParams(search), [search])
-  /**
-   * Keep track of the current searchParams value so that updater functions can
-   * use the current value rather than value from previous render.
-   */
-  const searchParamsRef = useRef(searchParams)
 
   useEffect(() => {
     hasNavigatedRef.current = false
-    /**
-     * Each render, sync the ref with the current state value.
-     * This is necessary in case search params has changed via some source other
-     * than this hook (e.g., browser navigation).
-     */
-    searchParamsRef.current = searchParams
   })
 
-  const setSearchParams: SetSearchParams = useCallback(
-    nextValue => {
-      const newParams =
-        typeof nextValue === "function" ?
-          nextValue(searchParamsRef.current) :
-          nextValue
-
-      searchParamsRef.current = newParams
-
-      if (hasNavigatedRef.current) {
-        router.replace(`${pathname}?${newParams}`, { scroll: false })
-      } else {
-        router.push(`${pathname}?${newParams}`, { scroll: false })
```

Comment:
The original implementation here was based on an earlier version specific to react router, which worked fine—for react router—but was a but buggy for NextJS.

[Apparently](https://github.com/vercel/next.js/issues/29240#issuecomment-923428999) in NextJS, multiple synchronous calls to `push` **or `replace`** are flattened into the most recent call. In this function, the first call was always a push, and the subsequent calls are always replace.

So if `setSearchParams` was called 2+ times, the "flatted call" would always be a `replace`, meaning no history stack.

**We could have fixed the bug by removing `replace` and always using `push`**. (Different from react-router).